runtime/parachains/src/inclusion/mod.rs

```diff
@@ -791,6 +950,92 @@ impl<T: Config> Pallet<T> {
 			)
 	}
 
+	/// Check that all the upward messages sent by a candidate pass the acceptance criteria. Returns
+	/// false, if any of the messages doesn't pass.
+	// FAIL-CI: TODO update docs
+	pub(crate) fn check_upward_messages(
+		config: &HostConfiguration<T::BlockNumber>,
+		para: ParaId,
+		upward_messages: &[UpwardMessage],
+	) -> Result<(), UmpAcceptanceCheckErr> {
+		// Cannot send UMP messages while off-boarding.
+		if <paras::Pallet<T>>::is_offboarding(para) {
+			ensure!(upward_messages.is_empty(), UmpAcceptanceCheckErr::IsOffboarding);
+		}
+
+		let additional_msgs = upward_messages.len();
+		if additional_msgs > config.max_upward_message_num_per_candidate as usize {
+			return Err(UmpAcceptanceCheckErr::MoreMessagesThanPermitted {
+				sent: additional_msgs as u32,
+				permitted: config.max_upward_message_num_per_candidate,
+			})
+		}
+
+		let fp = T::MessageQueue::footprint(AggregateMessageOrigin::Ump(para));
+		let (para_queue_count, mut para_queue_size) = (fp.count, fp.size);
+
+		if para_queue_count
+			.checked_add(additional_msgs as u64)
+			.map(|new| new > config.max_upward_queue_count as u64)
+			.unwrap_or(true)
+		{
+			return Err(UmpAcceptanceCheckErr::CapacityExceeded {
+				count: para_queue_count.saturating_add(additional_msgs as u64),
+				limit: config.max_upward_queue_count as u64,
+			})
+		}
+
+		for (idx, msg) in upward_messages.into_iter().enumerate() {
+			let msg_size = msg.len();
+			if msg_size > config.max_upward_message_size as usize {
+				return Err(UmpAcceptanceCheckErr::MessageSize {
+					idx: idx as u32,
+					msg_size: msg_size as u32,
+					max_size: config.max_upward_message_size,
+				})
+			}
+			// make sure that the queue is not overfilled.
+			// we do it here only once since returning false invalidates the whole relay-chain block.
+			if para_queue_size
+				.checked_add(msg_size as u64)
+				.map(|new| new > config.max_upward_queue_size as u64)
+				.unwrap_or(true)
```

Comment:
Would it make more sense to use `saturating_add` here instead of `checked_add`? I must admit that this `unwrap_or` here surprised me for a moment...

Reply:
What do you mean? It returns an error on overflow instead of saturating.  

Although currently it never will saturate, since it casts up from an `u32`. Is that what you meant @KiChjang?

Reply:
I wasn't expecting to see an `unwrap_or` here as I didn't expect seeing a `checked_add`, because it would seem to me that `saturating_add` makes more sense to use here rather than a `checked_add`, since you eventually unwrap the option anyway.